codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedCollectionGenerator.kt

```diff
@@ -0,0 +1,229 @@
+import software.amazon.smithy.rust.codegen.core.rustlang.RustMetadata
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.Visibility
+import software.amazon.smithy.rust.codegen.core.rustlang.docs
+import software.amazon.smithy.rust.codegen.core.rustlang.documentShape
+import software.amazon.smithy.rust.codegen.core.rustlang.join
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.server.smithy.PubCrateConstraintViolationSymbolProvider
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
+import software.amazon.smithy.rust.codegen.server.smithy.supportedCollectionConstraintTraits
+import software.amazon.smithy.rust.codegen.server.smithy.validationErrorMessage
+
+/**
+ * [ConstrainedCollectionGenerator] generates a wrapper tuple newtype holding a constrained `std::vec::Vec`.
+ * This type can be built from unconstrained values, yielding a `ConstraintViolation` when the input does not satisfy
+ * the constraints.
+ *
+ * The [`length`] and [`uniqueItems`] traits are the only constraint traits applicable to list shapes.
+ * The [`uniqueItems`] trait has not been implemented yet.
+ *
+ * If [unconstrainedSymbol] is provided, the `MaybeConstrained` trait is implemented for the constrained type, using the
+ * [unconstrainedSymbol]'s associated type as the associated type for the trait.
+ *
+ * [`length`]: https://smithy.io/2.0/spec/constraint-traits.html#length-trait
+ * [`uniqueItem`]: https://smithy.io/2.0/spec/constraint-traits.html#smithy-api-uniqueitems-trait
+ */
+class ConstrainedCollectionGenerator(
+    val codegenContext: ServerCodegenContext,
+    val writer: RustWriter,
+    val shape: CollectionShape,
+    private val constraintsInfo: List<TraitInfo>,
+    private val unconstrainedSymbol: Symbol? = null,
+) {
+    private val model = codegenContext.model
+    private val constrainedShapeSymbolProvider = codegenContext.constrainedShapeSymbolProvider
+    private val publicConstrainedTypes = codegenContext.settings.codegenConfig.publicConstrainedTypes
+    private val constraintViolationSymbolProvider =
+        with(codegenContext.constraintViolationSymbolProvider) {
+            if (publicConstrainedTypes) {
+                this
+            } else {
+                PubCrateConstraintViolationSymbolProvider(this)
+            }
+        }
+    private val symbolProvider = codegenContext.symbolProvider
+
+    fun render() {
+        assert(constraintsInfo.isNotEmpty()) {
+            "`ConstrainedCollectionGenerator` can only be invoked for constrained collections, but this shape was unconstrained"
```

Comment:
```suggestion
            "`ConstrainedCollectionGenerator` can only be invoked for constrained collections, but this shape was not constrained"
```

I've been using the term "unconstrained shapes" to refer to the unconstrained counterparts of constrained shapes.

Reply:
Done in 28a3397f527e4722d302743e2d5ef69800a21817.